### PR TITLE
feat: click outside directive

### LINKED ISSUE #1039: [FEATURE] Create click outside directive
**Describe how this feature will help us improve**
This directive will help us implement click outside method in components such as SfSelect or SfColorPicker.

**Additional context**
Name proposal for the directive: `v-click-out` / `v-close-out` / `v-closeable`

### PR DESCRIPTION
**Pull request by AdamPawlinski**
# Related issue
Closes #1039

# Scope of work
Closing component directive after clicking outside prepared. 
Directive added on SfSelect, SfDropdown, SfModal and SfSidebar.

# Checklist

- [x] No commented blocks of code left
- [ ] Run tests and docs
- [x] Self code-reviewed

### CODE REVIEW

File: packages/vue/src/components/molecules/SfSelect/SfSelect.vue [hunk 1]
```diff
@@ -13,11 +13,11 @@
     }"
     class="sf-select"
     @click="toggle($event)"
-    @blur="closeHandler"
     @keyup.space="openHandler"
     @keyup.up="move(-1)"
     @keyup.down="move(1)"
     @keyup.enter="enter($event)"
+    v-clickOutside="closeHandler"
```
Comment (martaradziszewska) [inline packages/vue/src/components/molecules/SfSelect/SfSelect.vue:20]: we can use a hyphenated name here like `v-click-outside`

File: packages/vue/src/utilities/directives/click-outside-directive.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+let outsideClickHandler;
```
[the remark this answers is not in the document]
Comment (AdamPawlinski) [reply packages/vue/src/utilities/directives/click-outside-directive.js]: Okay I fixed that.

File: packages/vue/src/utilities/directives/click-outside-directive.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+export const clickOutside = {
+  bind(el, binding) {
+    outsideClickHandler = (e) => {
+      e.stopPropagation();
+      const callback = binding.value;
```
[the remark this answers is not in the document]
Comment (AdamPawlinski) [reply packages/vue/src/utilities/directives/click-outside-directive.js]: changed to closeHandler, is that all right?

File: packages/vue/src/components/molecules/SfSelect/SfSelect.vue (hunk 1 above)
Comment (AdamPawlinski) [reply packages/vue/src/components/molecules/SfSelect/SfSelect.vue]: changed

**Comment by AdamPawlinski** [comment]
I made fixes Please take a look.

**Comment by martaradziszewska** [comment]
@AdamPawlinski could you try to implement this on SfDropdown? I am having some issues with implementation. Thanks!

**Comment by AdamPawlinski** [comment]
OK I'll try.

**Comment by AdamPawlinski** [comment]
Directive adjusted to dropdown and added to modal and sidebar. Please take a look.

**Comment by martaradziszewska** [comment]
@AdamPawlinski please resolve conflicts :)

**Comment by AdamPawlinski** [comment]
Conflicts resolved and few fixes made.

**Comment by martaradziszewska** [comment]
@AdamPawlinski I think for now the best solution is to change the event to `mouseup`.  In this case, the close outside will be invoked on click release and won't interfere with the button which invoked the open event. Wdyt? @pspaczek Please look at this :)

**Comment by AdamPawlinski** [comment]
@martaradziszewska great idea. I added exclude solution on select and I'm wondering if there could be some prop which could disable directive and point the elements not closing this elements at the same time. Sth like elementsToExclude which could be an array: 

-  when undefined the directive will be disabled, 

- when it consists of some elements the directive will be on 

-  when empty it will work as default.  
and the logic could be in helper which will be added to closing method in every component where is needed. Wdyt? We can discuss it later.

File: packages/vue/src/components/molecules/SfSelect/SfSelect.vue [hunk 4]
```diff
@@ -76,12 +78,13 @@ import SfSelectOption from "./_internal/SfSelectOption.vue";
 import SfChevron from "../../atoms/SfChevron/SfChevron.vue";
 import SfButton from "../../atoms/SfButton/SfButton.vue";
 import SfOverlay from "../../atoms/SfOverlay/SfOverlay.vue";
-import { focus } from "../../../utilities/directives";
+import { focus } from "../../../utilities/directives/";
+import { clickOutside } from "../../../utilities/directives/";
```
Comment (pspaczek) [inline packages/vue/src/components/molecules/SfSelect/SfSelect.vue:82]:
```suggestion
import { clickOutside } from "../../../utilities/directives";
```
probably you should adjust your imports to my suggestion

Comment (pspaczek) [reply packages/vue/src/components/molecules/SfSelect/SfSelect.vue:82]: @AdamPawlinski

File: packages/vue/src/components/molecules/SfSelect/SfSelect.vue [hunk 5]
```diff
@@ -22,6 +23,7 @@
       <!-- eslint-disable-next-line vue/no-v-html -->
       <div
         id="sfSelect"
+        ref="button"
```
Comment (pspaczek) [inline packages/vue/src/components/molecules/SfSelect/SfSelect.vue:26]: probably you can remove this ref

File: packages/vue/src/components/molecules/SfSelect/SfSelect.vue [hunk 6]
```diff
@@ -239,6 +242,8 @@ export default {
         this.disabled
       ) {
         return;
+      } else if (this.$refs["button"].contains(event.target) && this.open) {
+        this.open = false;
```
Comment (pspaczek) [inline packages/vue/src/components/molecules/SfSelect/SfSelect.vue:246]: it's still usefull? 
```suggestion
      } else if (this.$refs["button"].contains(event.target) && this.open) {
        this.open = false;
```

File: packages/vue/src/components/molecules/SfSelect/SfSelect.vue [hunk 7]
```diff
@@ -1,5 +1,7 @@
 <template>
   <div
+    v-click-outside="closeHandler"
+    :aria-expanded="open ? 'true' : 'false'"
```
Comment (pspaczek) [inline packages/vue/src/components/molecules/SfSelect/SfSelect.vue:4]:
```suggestion
    :aria-expanded="open.toString()"
```
wdyt about this solution ;)?

File: packages/vue/src/components/molecules/SfDropdown/SfDropdown.stories.js [hunk 8]
```diff
@@ -34,12 +34,13 @@ storiesOf("Molecules|Dropdown", module)
           "Add to wishlist",
           "Share",
         ],
+        notClosingEl: ["button"],
```
Comment (pspaczek) [inline packages/vue/src/components/molecules/SfDropdown/SfDropdown.stories.js:37]: is still useful?

**Pull request merged**